sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,821 @@
+
+    Subclasses of this class compute pairwise distances between a set of
+    vectors (rows) X and another set of vectors (rows) Y and apply a
+    reduction on top. The reduction takes a matrix of pairwise distances
+    between rows of X and Y as input and outputs an aggregate data-structure
+    for each row of X. The aggregate values are typically smaller than the number
+    of rows in Y, hence the term reduction.
+
+    For computational reasons, it is interesting to perform the reduction on
+    the fly on chunks of rows of X and Y so as to keep intermediate
+    data-structures in CPU cache and avoid unnecessary round trips of large
+    distance arrays with the RAM that would otherwise severely degrade the
+    speed by making the overall processing memory-bound.
+
+    The base class provides the generic chunked parallelization template using
+    OpenMP loops (Cython prange), either on rows of X or rows of Y depending on
+    their respective sizes.
+
+    The subclasses are specialized for reduction.
+
+    The actual distance computation for a given pair of rows of X and Y are
+    delegated to format-specific subclasses of the DatasetsPair companion base
+    class.
+
+    Parameters
+    ----------
+    datasets_pair: DatasetsPair
+        The pair of dataset to use.
+
+    chunk_size: int, default=None,
+        The number of vectors per chunk. If None (default) looks-up in
+        scikit-learn configuration for `pairwise_dist_chunk_size`,
+        and use 256 if it is not set.
+
+    n_threads: int, default=None
+        The number of OpenMP threads to use for the reduction.
+        Parallelism is done on chunks and the sharding of chunks
+        depends on the `strategy` set on :method:`~PairwiseDistancesReduction.compute`.
+
+        See _openmp_effective_n_threads, for details about
+        the specification of n_threads.
+
+    strategy : str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default=None
+        The chunking strategy defining which dataset parallelization are made on.
+
+        Strategies differs on the dispatching they use for chunks on threads:
+
+          - 'parallel_on_X' dispatches chunks of X uniformly on threads.
+          Each thread then iterates on all the chunks of Y. This strategy is
+          embarrassingly parallel and comes with no datastructures synchronisation.
+
+          - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
+          Each thread then iterates on all the chunks of X. This strategy is
+          embarrassingly parallel but uses intermediate datastructures
+          synchronisation.
+
+          - 'auto' relies on a simple heuristic to choose between
+          'parallel_on_X' and 'parallel_on_Y'.
+
+          - None (default) looks-up in scikit-learn configuration for
+          `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+    """
+
+    cdef:
+        readonly DatasetsPair datasets_pair
```

Comment:
Why are all these defined as class/static variables instead of instance variables (set in __init__)?

Reply:
For the sake of clarity (because I do not know what you are referring to using "these"), I will be verbose:

All the definition under `cdef`-scope are [instance attributes definitions](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#introduction) not class attributes definitions.

Moreover,
```
cdef readonly DatasetsPair datasets_pair
```
is used to have [this attribute accessible from Python scopes](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#static-attributes).

Does that answer your question?

Reply:
Yes. Also, I was confused. It's fine as it is.